src/osd/PrimaryLogPG.cc

```diff
@@ -4495,42 +4495,37 @@ void PrimaryLogPG::do_scan(
 	return;
       }
 
-      BackfillInterval bi;
-      bi.begin = m->begin;
       // No need to flush, there won't be any in progress writes occuring
       // past m->begin
-      scan_range(
+      BackfillInterval bi = scan_range(
 	cct->_conf->osd_backfill_scan_min,
 	cct->_conf->osd_backfill_scan_max,
-	&bi,
+	m->begin,
+	info.last_update,
 	handle);
       MOSDPGScan *reply = new MOSDPGScan(
-	MOSDPGScan::OP_SCAN_DIGEST,
+	MOSDPGScan::OP_SCAN_DIGEST_REPLY,
 	pg_whoami,
 	get_osdmap_epoch(), m->query_epoch,
-	spg_t(info.pgid.pgid, get_primary().shard), bi.begin, bi.end);
-      encode(bi.objects, reply->get_data());
+	spg_t(info.pgid.pgid, get_primary().shard),bi.get_begin(), bi.get_end(),
+	bi.get_interval_version());
+      encode(bi.get_objects(), reply->get_data());
       osd->send_message_osd_cluster(reply, m->get_connection());
     }
     break;
 
-  case MOSDPGScan::OP_SCAN_DIGEST:
+  case MOSDPGScan::OP_SCAN_DIGEST_REPLY:
     {
       pg_shard_t from = m->from;
 
       // Check that from is in backfill_targets vector
       ceph_assert(is_backfill_target(from));
+      BackfillInterval bi(m->begin, m->end);
+      bi.populate(m->get_data(), m->version);
```

Comment:
Seems like it would be cleaner for MOSDScanDigestReply to have a get_interval() method which returns a fully constructed BackfillInterval.